src/shims/unix/mem.rs

```diff
@@ -0,0 +1,255 @@
+
+        let (prov, offset) = pointer.into_parts();
+        let alloc_id = match prov {
+            Some(Provenance::Concrete { alloc_id, sb: _ }) => alloc_id,
+            _ => panic!("realloc should return a pointer with provenance"),
+        };
+        let start =
+            intptrcast::GlobalStateInner::rel_ptr_to_addr(this, Pointer::new(alloc_id, offset));
+        this.machine
+            .mappings
+            .all_memory
+            .iter_mut(Size::from_bytes(start), Size::from_bytes(new_size))
+            .take(1) // "an existing mapping" meaning only one mapping starting at old_address?
+            .for_each(|(_offset, map)| {
+                if let Some(m) = map {
+                    m.alloc_id = alloc_id;
+                }
+            });
+
+        Ok(pointer)
+    }
+
+    fn munmap(
+        &mut self,
+        addr: &OpTy<'tcx, Provenance>,
+        length: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let addr = this.read_scalar(addr)?.to_machine_usize(this)?;
+        let length = this.read_scalar(length)?.to_machine_usize(this)?;
+
+        // The address addr must be a multiple of the page size (but length need not be).
+        #[allow(clippy::integer_arithmetic)] // PAGE_SIZE is nonzero
+        if addr % PAGE_SIZE != 0 {
+            this.set_last_error(Scalar::from_i32(this.eval_libc_i32("EINVAL")?))?;
+            return Ok(-1);
+        }
+
+        // All pages containing a part of the indicated range are unmapped.
+        let length = round_up_to_multiple_of_page_size(length).unwrap_or(u64::MAX);
+
+        for (_offset, map) in this
+            .machine
+            .mappings
+            .all_memory
+            .iter_mut(Size::from_bytes(addr), Size::from_bytes(length))
+        {
+            *map = None;
+        }
+        Ok(0)
+    }
+
+    fn mprotect(
+        &mut self,
+        addr: &OpTy<'tcx, Provenance>,
+        length: &OpTy<'tcx, Provenance>,
+        prot: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+        let addr = this.read_scalar(addr)?.to_machine_usize(this)?;
+        let length = this.read_scalar(length)?.to_machine_usize(this)?;
+        let prot = this.read_scalar(prot)?.to_i32()?;
+
+        let prot_read = this.eval_libc_i32("PROT_READ")?;
+        let prot_write = this.eval_libc_i32("PROT_WRITE")?;
+
+        // All pages containing a part of the indicated range are modified
+        let length = round_up_to_multiple_of_page_size(length).unwrap_or(u64::MAX);
+
+        for (_offset, map) in this
+            .machine
+            .mappings
+            .all_memory
+            .iter_mut(Size::from_bytes(addr), Size::from_bytes(length))
+        {
+            if let Some(m) = map {
+                m.can_read = prot & prot_read > 0;
+                m.can_write = prot & prot_write > 0;
+            }
+        }
```

Comment:
So this will just silently do nothing if the memory in the given region is not in any mapping (e.g. if it was allocated in another way)?